FirebaseStorage/Sources/StorageDownloadTask.swift

```diff
@@ -93,99 +97,98 @@ open class StorageDownloadTask: StorageObservableTask, StorageTaskManagement {
+    } else {
+      let fetcherService = await reference.storage.fetcherService.fetcherService(reference.storage)
+
+      fetcher = fetcherService.fetcher(with: request)
+      fetcher.comment = "Starting DownloadTask"
+    }
+    fetcher.maxRetryInterval = reference.storage.maxDownloadRetryInterval
+
+    if let fileURL {
+      // Handle file downloads
+      fetcher.destinationFileURL = fileURL
+      fetcher.downloadProgressBlock = { [weak self] (bytesWritten: Int64,
+                                                     totalBytesWritten: Int64,
+                                                     totalBytesExpectedToWrite: Int64) in
+          guard let self = self else { return }
+          self.state = .progress
+          self.progress.completedUnitCount = totalBytesWritten
+          self.progress.totalUnitCount = totalBytesExpectedToWrite
+          self.fire(for: .progress, snapshot: self.snapshot)
+          self.state = .running
+      }
+    } else {
+      // Handle data downloads
+      fetcher.receivedProgressBlock = { [weak self] (bytesWritten: Int64,
+                                                     totalBytesWritten: Int64) in
+          guard let self = self else { return }
+          self.state = .progress
+          self.progress.completedUnitCount = totalBytesWritten
+          if let totalLength = self.fetcher?.response?.expectedContentLength {
+            self.progress.totalUnitCount = totalLength
           }
-          self.fire(for: .success, snapshot: self.snapshot)
-        } catch {
           self.fire(for: .progress, snapshot: self.snapshot)
-          self.state = .failed
-          self.error = StorageErrorCode.error(
-            withServerError: error as NSError,
-            ref: self.reference
-          )
-          self.fire(for: .failure, snapshot: self.snapshot)
-        }
-        self.removeAllObservers()
+          self.state = .running
+      }
+    }
+    self.fetcher = fetcher
+    state = .running
+    do {
+      let data = try await self.fetcher?.beginFetch()
+      // Fire last progress updates
+      fire(for: .progress, snapshot: snapshot)
+
+      // Download completed successfully, fire completion callbacks
+      state = .success
+      if let data {
+        downloadData = data
       }
+      fire(for: .success, snapshot: snapshot)
+    } catch {
+      fire(for: .progress, snapshot: snapshot)
+      state = .failed
+      self.error = StorageErrorCode.error(
+        withServerError: error as NSError,
+        ref: reference
+      )
+      fire(for: .failure, snapshot: snapshot)
     }
+    removeAllObservers()
   }
 
   func cancel(withError error: NSError) {
     dispatchQueue.async { [weak self] in
       guard let self = self else { return }
       self.state = .cancelled
       self.fetcher?.stopFetching()
+      print("stopped fetching")
       self.error = error
       self.fire(for: .failure, snapshot: self.snapshot)
+      self.removeAllObservers()
```

Comment:
The integration tests had a success after cancel event. I believe this was a bug before.